ci: Publish ebpf builder helper image

mqasimsarfraz opened the pull request:
Fixes: #1420

It uses the same idea of #1472 and will only:
- Build images if changes are detected
- Publish images if variable `PUSH_HELPERS = ENABLE_PUSH_HELPERS` is defined (it helps avoid issues with external contributors PR)
- Publish images for each release

## Testing done

- Fork: A [test commit](https://github.com/mqasimsarfraz/inspektor-gadget/commit/33523afbb00c11be79102946f0d876bd033b56c8) result in helper [image being built](https://github.com/mqasimsarfraz/inspektor-gadget/actions/runs/6482505619/job/17602162537) and used [here](https://github.com/mqasimsarfraz/inspektor-gadget/actions/runs/6482505619/job/17602170043#step:4:1).
- Release: A test release (`v0.10.0`) publish the tags [here](https://github.com/mqasimsarfraz/inspektor-gadget/pkgs/container/ebpf-builder/versions).

## Todo

- [x] Claim ownership of https://github.com/orgs/inspektor-gadget/packages/container/package/ebpf-builder since it belongs to Inspektor Gadget `org` (or someone who created it) not `repo`. 

## Follow up
- [ ] Add support for arm64
- [ ]  Use the published release image [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/cmd/common/image/build.go#L45). (Should we? Separate PR?)

alban commented on .github/workflows/inspektor-gadget.yml:
```diff
@@ -657,9 +660,15 @@ jobs:
+          - name: "ebpf-builder"
+            context: "/home/runner/work/inspektor-gadget/inspektor-gadget"
+            dockerfile: "Dockerfiles/ebpf-builder.Dockerfile"
+            platform: "linux/amd64"
```
Does it mean `ig image build` would not work on linux/arm64?

mqasimsarfraz replied:
I saw the [existing image](https://github.com/inspektor-gadget/inspektor-gadget/pkgs/container/ebpf-builder) was also only built for `linux/amd64` so I just used that but yes that would mean `ig image build` won't work on `linux/arm64` with builder image. Do you know if it fine to build it for `linux/arm64` ?

mqasimsarfraz replied:
I tried building it for `arm64` and running into following [error](https://github.com/mqasimsarfraz/inspektor-gadget/actions/runs/6493130779/job/17633458645#step:8:1181):
```
#9 33.43 E: Package 'gcc-multilib' has no installation candidate
#9 ERROR: process "/dev/.buildkit_qemu_emulator /bin/sh -c apt-get update \t&& apt-get install -y gcc-multilib lsb-release wget software-properties-common gnupg" did not complete successfully: exit code: 100
```

mqasimsarfraz replied:
I see, I opened a separate PR (it will also test if image is built fine for PR) for this, https://github.com/inspektor-gadget/inspektor-gadget/pull/2139 so we can track `arm64` alongside.

mqasimsarfraz replied:
It failed while building/pushing gadgets: https://github.com/inspektor-gadget/inspektor-gadget/actions/runs/6495111185/job/17642715987

alban reviewed (commented):
> Claim ownership of https://github.com/orgs/inspektor-gadget/packages/container/package/ebpf-builder since it belongs to Inspektor Gadget org (or someone who created it) not repo.

I did it manually. But you can add the label in `Dockerfiles/ebpf-builder.Dockerfile`:

```
LABEL org.opencontainers.image.source https://github.com/inspektor-gadget/inspektor-gadget
```

(although the `LABEL` keyword in the Dockerfile might not work for multi-arch images)

> Use the published release image [here](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/cmd/common/image/build.go#L45). (Should we? Separate PR?)

I think we should with the following rule:
- If VERSION is a release, use that tag by default
- If not, use `latest` by default

What do you think?

mauriciovasquezbernal commented:
> I think we should with the following rule:
> 
>     * If VERSION is a release, use that tag by default
> 
>     * If not, use `latest` by default

I think it if the image is modified by a PR, then the built image for that specific PR should be used, otherwise it'll be impossible to implement PRs that modify the image and the gadgets (as https://github.com/inspektor-gadget/inspektor-gadget/pull/2107) without having a "REMOVE BEFORE MERGE" commit pointing the image to something else.

mauriciovasquezbernal commented on .github/workflows/inspektor-gadget.yml:
```diff
@@ -677,25 +677,25 @@ jobs:
           pattern: ${{ matrix.image.filter-pattern }}
     - name: Set up Docker Buildx
-      if: steps.filter.outputs.pattern == 'true'
+      if: steps.filter.outputs.pattern == 'true' || startsWith(github.ref, 'refs/tags/v')
```
Would it make sense to define a new `should-skip` step with this logic and then other steps will only check a single output from the `should-skip` step?

mauriciovasquezbernal reviewed (changes requested):
Looks good mostly, only blocking comment on missing rules on pattern.

IMO we can handle arm64 in another PR. 

Btw, I rebased https://github.com/inspektor-gadget/inspektor-gadget/pulls on top of it so we can check with a real example if it's working fine :D

mqasimsarfraz commented on .github/workflows/inspektor-gadget.yml:
```diff
@@ -657,9 +660,15 @@ jobs:
+            context: "/home/runner/work/inspektor-gadget/inspektor-gadget"
+            dockerfile: "Dockerfiles/ebpf-builder.Dockerfile"
+            platform: "linux/amd64"
+            filter-pattern: "Dockerfiles/ebpf-builder.Dockerfile"
```
Done.

mqasimsarfraz commented:
> Btw, I rebased https://github.com/inspektor-gadget/inspektor-gadget/pulls on top of it so we can check with a real example if it's working fine :D

I see it failed because of missing patterns, I updated the list and now it [looking better](https://github.com/mqasimsarfraz/inspektor-gadget/actions/runs/6493511261) with [PR on fork](https://github.com/mqasimsarfraz/inspektor-gadget/pull/37)!

mqasimsarfraz commented:
I merged this one and we can track `arm64` in https://github.com/inspektor-gadget/inspektor-gadget/pull/2139
